### PR TITLE
Fix tests broken by the element-wise pair assignment

### PR DESCRIPTION
**Pull request by Bastacyclop**
Hacks the continuation translation of foreign functions to avoid calling the function multiple times when triggering a pair assignment.
Removes the acceptor translation of foreign functions, as it should not be needed?

It is open to discussion whether this fix should be merged or if we prefer waiting for a better solution.
Related to #58

### CODE REVIEW

File: src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala [hunk 1]
```diff
@@ -34,23 +34,7 @@ final case class ForeignFunction(funDecl: ForeignFunction.Declaration,
...
-          con(arg)(λ(expT(inT, read))(e =>
-            A :=| outT | ForeignFunction(funDecl, inTs :+ inT, outT, exps :+ e)))
-        // with a `tail` of arguments left, recurse
-        case Seq((arg, inT), tail@_*) =>
-          con(arg)(λ(expT(inT, read))(e => recurse(tail, exps :+ e, inTs :+ inT)))
-      }
-    }
-
-    recurse(args zip inTs, Seq(), Seq())
+    ???
```
Comment (michel-steuwer) [inline src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala:37]: Can you explain why it is required (or a good idea) to remove the acceptor translation?

File: src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala [hunk 2]
```diff
@@ -64,7 +48,24 @@ final case class ForeignFunction(funDecl: ForeignFunction.Declaration,
         // with only one argument left to process return the assignment of the function call
         case Seq( (arg, inT) ) =>
           con(arg)(λ(expT(inT, read))(e =>
-            C( ForeignFunction(funDecl, inTs :+ inT, outT, exps :+ e) )) )
+            outT match {
+              // TODO: this is an ugly fix to avoid calling the function multiple times
+              //  for pair assignment
```
Comment (michel-steuwer) [inline src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala:53]: Could we have an explicit link to the issue discussing this in the comment here?

File: src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala (hunk 1 above)
Comment (Bastacyclop) [reply src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala:37]: the `ForeignFunction` primitive is `read`, and so should only trigger continuation translation I believe.

File: src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala]: done

**Comment by michel-steuwer** [comment]
:+1:

**Review by michel-steuwer — APPROVED** [review]

**Comment by michel-steuwer** [comment]
I suggest that we merge this while keeping the issue open.
It fixes two tests and doesn't seem to introduce harm to the rest of the code base.

File: src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala (hunk 1 above)
Comment (fedepiz) [reply src/main/scala/shine/DPIA/FunctionalPrimitives/ForeignFunction.scala:37]: I may be wrong on this, but doesn't a top-level `read` primitive that returns a simple type like `f32` still trigger the acceptor translation?

I think I am facing exactly this in my current implementation of dependent pair match, which in one example is of type `exp[f32, read]`, and triggers the acceptor translation?

Again, I am rusty in `shine`, but it's worth a check

**Review by fedepiz — APPROVED** [review]
Looks good

**Pull request merged**